### PR TITLE
Improve `ExGram.Dispatcher` and related modules

### PR DESCRIPTION
**Pull request by sega-yarkin**
Refactor and simplify `Dispatcher`, `Bot`, and `Bot.Supervisor` modules.

### CODE REVIEW

**Comment by sega-yarkin** [comment]
Hello @rockneurotiko! What do you think about this update? If it seems good to merge I will add more changes, like function specs, also maybe some tests. Will happy to discuss changes.

**Comment by rockneurotiko** [comment]
Sorry for the delay, I'll check the PR when I have some free time!!

File: lib/ex_gram/dispatcher.ex [hunk 1]
```diff
@@ -224,57 +221,50 @@ defmodule ExGram.Dispatcher do
...
-        |> ExGram.Dsl.send_answers()
-        |> handle_responses(error_handler)
+  defp apply_middlewares(%Cnt{middlewares: [_ | rest]} = cnt) do
+    apply_middlewares(%Cnt{cnt | middlewares: rest})
+  end
 
-      _ ->
-        :noop
+  defp call_handler({module, method}, info, cnt, error_handler) do
+    with %Cnt{} = cnt <- apply(module, method, [info, cnt]) do
```
Comment (rockneurotiko) [inline lib/ex_gram/dispatcher.ex:252]: Since this is just one use of `<-` on the with, let's use a case here

**Review by rockneurotiko — APPROVED** [review]
I really like the changes, much cleaner and with modern semantics instead of some old ones that the project still had.

Did you tried it in some bot? I don't see anything that would break it. I will try it too

**Comment by sega-yarkin** [comment]
@rockneurotiko Thanks for the feedback!

> Did you tried it in some bot?

I've recently started a side project where I wanna use TG bot, not much time for it, but I do my development using this branch, and so far it looks good.

**Comment by sega-yarkin** [comment]
Hi @rockneurotiko! I've added a couple of things into PR, also tested more with my setup, and basic things like receiving and sending back messages works fine. Can you please check the changes and if all looks good to you, I will merge it? Thanks

**Comment by sega-yarkin** [comment]
Hi @rockneurotiko! Have you gotten a chance to test this change? I was awaiting for a response from you and this PR fall off my focus. I've just re-based it on top of latest changes. Looks like someone is already awaiting on this PR, so if you find I can merge it now. TIA.

File: lib/ex_gram/bot/supervisor.ex [hunk 2]
```diff
@@ -50,56 +50,39 @@ defmodule ExGram.Bot.Supervisor do
...
 
-  defp maybe_fetch_bot(username, _token) when is_binary(username),
+  defp get_bot_info(username, _token) when is_binary(username),
     do: %ExGram.Model.User{username: username, is_bot: true}
 
-  defp maybe_fetch_bot(_username, token) do
+  defp get_bot_info(_username, token) do
     case ExGram.get_me(token: token) do
       {:ok, bot} -> bot
       _ -> nil
```
Comment (sega-yarkin) [inline lib/ex_gram/bot/supervisor.ex:73]: It feels like `get_me` shouldn't fail (except network or token issue) and should return `ExGram.Model.User` all the times, if so we can make `bot_info` required for `Dispatcher`.

**Comment by rockneurotiko** [comment]
Sorry for the delay! Thanks for the PR ❤️

**Pull request merged**